src/west/commands/project.py

```diff
@@ -491,31 +491,45 @@ def _projects(args, listed_must_be_cloned=True):
         # No projects specified. Return all projects.
         return projects
 
-    # Got a list of projects on the command line. First, check that they exist
-    # in the manifest.
-
-    project_names = [project.name for project in projects]
-    nonexistent = set(args.projects) - set(project_names)
-    if nonexistent:
-        log.die('Unknown project{} {} (available projects: {})'
-                .format('s' if len(nonexistent) > 1 else '',
-                        ', '.join(nonexistent),
-                        ', '.join(project_names)))
-
-    # Return the projects in the order they were listed
+    # Listed but missing projects. Used for error reporting.
+    missing_projects = []
+
     res = []
-    for name in args.projects:
+    for project_arg in args.projects:
         for project in projects:
-            if project.name == name:
+            if project.name == project_arg:
+                # The argument is a project name
                 res.append(project)
                 break
+        else:
+            # The argument is not a project name. See if it is a project
+            # (sub)path.
+            for project in projects:
+                # The startswith() means we also detect subdirectories of
+                # project repositories. Giving a plain file in the repo will
+                # work here too, but that probably doesn't hurt.
+                if os.path.realpath(project_arg).startswith(
```

Comment:
The project with the "closest" match (i.e. longest common normalized subpath) should win in case there are multiple, rather than halting at the first potential match.

This is important for emulating git submodules with west. You can do that in repo by setting a project path to the correct submodule location and the right revision (it's in fact another use case for having the manifest revision be a plain SHA).

MCUboot is again a real world use case for this, as it has mbedtls as a submodule.

If I have a `project` and a `project/submodule`, and I run `west {whatever} .` from `submodule`, it seems like `project` would match here if it was listed first in the manifest.

Reply:
Fixed by sorting the projects with the longest `abspath` first.